### ISSUE TITLE
Support inferring dict literals to be a union of TypedDicts

### ISSUE THREAD

**Issue by RobertCraigie** [python/mypy#13274]
**Feature**

```py
from __future__ import annotations
from typing import TypedDict

class A(TypedDict, total=False):
    name: str

class B(TypedDict, total=False):
    name: str

def foo(data: A | B) -> None:
    ...

# this should not report an error
foo({"name": "Robert"})  # error: Type of TypedDict is ambiguous, could be any of ("A", "B")
```

**Pitch**

While the code example provided is nonsensical, I ran into this error using real world types. In the more abstract sense, inferring the `{"name": "foo"}` expression to be `A | B` should be supported as it is type safe. On top of that, Pyright also supports this pattern and infers the expression correctly.

There is also a "bug" with the check in that it doesn't respect nested types, consider this case:

```py
from __future__ import annotations
from typing import TypedDict

class A(TypedDict, total=False):
    foo: C

class B(TypedDict, total=False):
    foo: D

class C(TypedDict, total=False):
    c: str

class D(TypedDict, total=False):
    d: str

def foo(data: A | B) -> None:
    ...

foo({"foo": {"c": "foo"}})  # error: Type of TypedDict is ambiguous, could be any of ("A", "B")
```

In this case, there is no ambiguity, the expression can be safely inferred to be of type `A`.

**Comment by rattrayalex** [comment]
Possibly related: https://github.com/python/mypy/issues/8533

**Comment by jepperaskdk** [comment]
I'm also getting this (python3.6, mypy=0.971)
```
from typing import Union
from typing_extensions import Literal, TypedDict


class A(TypedDict):
    type: Literal['a']
    value: bool


class B(TypedDict):
    type: Literal['b']
    value: str


Response = Union[A, B]


def method(message: Response) -> None:
    ...


method({'type': 'a', 'value': True})
# error: Type of TypedDict is ambiguous, could be any of ("A", "B")
# error: Argument 1 to "method" has incompatible type "Dict[str, object]"; expected "Union[A, B]"


method({'type': 'b', 'value': "abc"})
# error: Type of TypedDict is ambiguous, could be any of ("A", "B")
# error: Argument 1 to "method" has incompatible type "Dict[str, str]"; expected "Union[A, B]"
```
